clouddriver-saga/src/test/kotlin/com/netflix/spinnaker/clouddriver/saga/examples/BranchingExampleTest.kt

```diff
@@ -0,0 +1,133 @@
+import com.netflix.spinnaker.clouddriver.saga.models.Saga
+import dev.minutest.rootContext
+import strikt.api.expectThat
+import strikt.assertions.isEqualTo
+import java.util.function.Predicate
+
+/**
+ * This example shows how to do branching logic inside of Sagas.
+ */
+class BranchingExampleTest : AbstractSagaTest() {
+  fun tests() = rootContext<Fixture> {
+
+    context("branching logic") {
+      fixture { Fixture() }
+
+      val flow = SagaFlow()
+        .then(PrepareAction::class.java)
+        .then(TheThingAction::class.java)
+        .on(ShouldDoOptionalThings::class.java) {
+          it.then(AnOptionalThingAction::class.java)
+        }
+        .then(FinishAction::class.java)
+        .completionHandler(ThingsCompletedHandler::class.java)
+
+      test("branch skipped") {
+        expectThat(sagaService.applyBlocking<String>(flow, PrepareForThings("mytest", "id", false)))
+          .isEqualTo("not branch")
+      }
+
+      test("branch entered") {
+        expectThat(sagaService.applyBlocking<String>(flow, PrepareForThings("mytest", "id", true)))
+          .isEqualTo("branch")
+      }
+    }
+  }
+
+  private inner class Fixture : BaseSagaFixture() {
+    init {
+      registerBeans(
+        applicationContext,
+        ThingsCompletedHandler::class.java,
+        PrepareAction::class.java,
+        TheThingAction::class.java,
+        AnOptionalThingAction::class.java,
+        FinishAction::class.java,
+        ShouldDoOptionalThings::class.java
+      )
+    }
+  }
+
+  @JsonTypeName("prepareForThings")
+  class PrepareForThings(sagaName: String, sagaId: String, val doOptionalThings: Boolean) : SagaCommand(sagaName, sagaId)
+
+  @JsonTypeName("doTheThing")
+  class DoTheThing(sagaName: String, sagaId: String) : SagaCommand(sagaName, sagaId)
+
+  @JsonTypeName("doAnOptionalThing")
+  class DoAnOptionalThing(sagaName: String, sagaId: String) : SagaCommand(sagaName, sagaId)
+
+  @JsonTypeName("finishThings")
+  class FinishThings(sagaName: String, sagaId: String) : SagaCommand(sagaName, sagaId)
+
+  class PrepareAction : SagaAction<PrepareForThings> {
+    override fun apply(command: PrepareForThings, saga: Saga): SagaAction.Result {
+      return SagaAction.Result(DoTheThing(saga.name, saga.id), listOf())
+    }
+  }
+
+  class TheThingAction : SagaAction<DoTheThing> {
+    override fun apply(command: DoTheThing, saga: Saga): SagaAction.Result {
+      // TODO(rz): Add a condition predicate that just checks for whether or not the command exists at all instead?
+      return SagaAction.Result(
+        ManyCommands(
+          DoAnOptionalThing(saga.name, saga.id),
+          FinishThings(saga.name, saga.id
+        )),
+        listOf()
+      )
+    }
+  }
+
+  class AnOptionalThingAction : SagaAction<DoAnOptionalThing> {
+    override fun apply(command: DoAnOptionalThing, saga: Saga): SagaAction.Result {
+      return SagaAction.Result(FinishThings(saga.name, saga.id), listOf())
```

Comment:
sort of feels like the end of the conditional branch should be able to return SagaAction.Result.EMPTY because the main flow of execution is already saying it is going a Many of DoAnOptionalThing and FinishThings

Reply:
Yes. You're right, oops.